runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -1493,6 +1500,196 @@ J9::CodeGenerator::lowerTreeIfNeeded(
+      //
+      // into
+      //   +--------------------------------+
+      //   | treetop                        |
+      //   |   <array-reference>            |
+      //   | treetop                        |
+      //   |   <value-reference>            |
+      //   | ificmpeq  -->------------------*---------+
+      //   |   iand                         |         |
+      //   |     iloadi <isClassFlags>      |         |
+      //   |       aloadi <componentClass>  |         |
+      //   |         aloadi <vft-symbol>    |         |
+      //   |           <array-reference>    |         |
+      //   |     iconst J9ClassIsValueType  |         |
+      //   |   iconst 0                     |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //   | BBStart (Extension)            |         |
+      //   | NULLCHK                        |         |
+      //   |   Passthrough                  |         |
+      //   |     <value-reference>          |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //                   |                          |
+      //                   +--------------------------+
+      //                   |
+      //                   v
+      //   +--------------------------------+
+      //   | BBStart                        |
+      //   | ArrayStoreCHK                  |
+      //   |   astorei/awrtbari             |
+      //   |     aladd                      |
+      //   |       aload <array>            |
+      //   |       index-offset-calculation |
+      //   |     aload <value>              |
+      //   +--------------------------------+
+      //
+      TR::SymbolReference *vftSymRef = self()->comp()->getSymRefTab()->findOrCreateVftSymbolRef();
+      TR::SymbolReference *arrayCompSymRef = self()->comp()->getSymRefTab()->findOrCreateArrayComponentTypeSymbolRef();
+      TR::SymbolReference *classFlagsSymRef = self()->comp()->getSymRefTab()->findOrCreateClassFlagsSymbolRef();
+
+      TR::Node *vft = TR::Node::createWithSymRef(node, TR::aloadi, 1, anchoredArrayTT->getNode()->getFirstChild(), vftSymRef);
+      TR::Node *arrayCompClass = TR::Node::createWithSymRef(node, TR::aloadi, 1, vft, arrayCompSymRef);
+      TR::Node *loadClassFlags = TR::Node::createWithSymRef(node, TR::iloadi, 1, arrayCompClass, classFlagsSymRef);
+      TR::Node *isValueTypeNode = TR::Node::create(node, TR::iand, 2, loadClassFlags, TR::Node::iconst(node, J9ClassIsValueType));
+
+      TR::Node *ifNode = TR::Node::createif(TR::ificmpeq, isValueTypeNode, TR::Node::iconst(node, 0));
+      ifNode->copyByteCodeInfo(node);
+
+      TR::Node *passThru  = TR::Node::create(node, TR::PassThrough, 1, sourceChild);
+      TR::ResolvedMethodSymbol *currentMethod = self()->comp()->getMethodSymbol();
+
+      TR::Block *arrayStoreCheckBlock = prevBlock->splitPostGRA(tt, cfg);
```

Comment:
@hzongaro  Changes looks good to me, one small thing I noticed is that you are calling `splitPostGRA` before creating treetops for `NULLCHK` and `isValueTypeNode` test. For the purpose of this change it is fine as those new trees would **not introduce** any node that needs un-commoning, but if there were, you will introduce a bug where a node is live outside of its extended basic block, as block splitter walks treetops to identify nodes which would need uncommoning.

Reply:
@r30shah If you feel that would make the code easier to understand, and avoid any potential future bugs - especially if someone needs to model a future change on this - I would be happy to adjust this accordingly, either under this pull request or in a follow up PR to clean it up. . . .

Reply:
@hzongaro Sorry about this minor nitpick, I am fine with the way code is interacting with splitter currently but if you want to make changes, as you are creating nodes for ifNode and NULLCHK before splitting block, you can create and insert the treetop before `tt` and `anchoredSourceTT` and can split the block from `tt`.

Reply:
Thanks @hzongaro  for pointing out the issue with the change I suggested in previous comment,(Issue with the change I suggested is that it will introduce the regStore trees after the control flow trees before ending basic block. Which is not functionally correct). Taking my suggestion back, Changes looks completely fine to me.

Reply:
Thanks for reviewing, Rahil @r30shah!